### PR TITLE
ore: Make upload --force deregister AMI too

### PR DESCRIPTION
**Pull request by cgwalters**
This matches the semantic intended for coreos-assembler.

Closes: https://github.com/coreos/mantle/issues/1035

### CODE REVIEW

**Comment by cgwalters** [comment]
Only compile tested so far.

**Comment by bgilbert** [comment]
Per https://github.com/coreos/mantle/issues/1035#issuecomment-516547129, this should try to delete the underlying snapshot as well (otherwise we're leaking storage), but continue if that fails (since something else may also hold a reference to the snapshot).  Also I'd tend to agree that deletion should be handled by ore calling primitives, rather than as a special path in `createImage`.

**Comment by cgwalters** [comment]
Bigger picture, do we care about ore's existing behavior?  Any reason not to unconditionally do the equivalent of `--force` and optimize it for cosa?

**Comment by bgilbert** [comment]
I don't think there are compatibility constraints for ore (unlike plume) but unconditionally deleting a random image with the same name seems a bit antisocial.  What about failing by default if the image already exists (by calling `FindImage` from ore), unless the user specifies either e.g. `--reuse-existing` or `--overwrite-existing`?

**Comment by arithx** [comment]
As @bgilbert mentioned I'd prefer that we handle the `FindImage` call followed by a new `DeleteImage` (and `DeleteSnapshot`) call inside of `cmd/ore/aws/upload.go` directly rather than adding it inside of the `createImage` function.

In the case of `DeleteSnapshot` if it provides any error the behavior should be to warn & print the ID (there's no other easy way to fetch once the AMI is deleted as mentioned in https://github.com/coreos/mantle/issues/1035#issuecomment-516547129).

----

> What about failing by default if the image already exists (by calling FindImage from ore), unless the user specifies either e.g. --reuse-existing or --overwrite-existing?

I think this can lead to a can of worms, do these options refer to S3 objects, snapshots, or AMIs? I'd probably prefer to have `--force` & explicit flags like `--reuse-s3`, `--reuse-snapshot` & `--reuse-ami`.

We might want to punt this discussion to another issue / future PR so as to get this in faster.

----

Side note for the current code: the description for the `--force` flag needs to be updated to reflect the new functionality (github won't let me comment on lines you didn't change).

**Comment by cgwalters** [comment]
> but unconditionally deleting a random image with the same name seems a bit antisocial

Hardly random, you'd have to be in the same AWS account and using the same OS name right?

**Comment by bgilbert** [comment]
> Hardly random, you'd have to be in the same AWS account and using the same OS name right?

Well, yeah.  So the issue is shared dev (or prod?) accounts.

Note: I do have a preference here but I don't feel super-strongly about it.

**Comment by cgwalters** [comment]
With https://github.com/coreos/coreos-assembler/pull/693 the snapshot won't get lost.

And I think garbage collection is best implemented in cosa after we land https://github.com/coreos/coreos-assembler/issues/668

**Comment by bgilbert** [comment]
> With coreos/coreos-assembler#693 the snapshot won't get lost.

For uploads initiated by cosa, not those initiated by hand, which after all is the primary purpose of ore.  There's no point in deleting the AMI if we leave behind the part that costs money.

**Comment by cgwalters** [comment]
Thought about this some more and I think the real fix has to be https://github.com/coreos/coreos-assembler/issues/668 - reusing version numbers as we are doing now is just too prone to confusion and failure.

Anyways, I redid the patch per request since I think the existing status quo is really bad - it took me a lot of confusing debugging to realize that the AMI had been reused from a previous failed build.

**Comment by cgwalters** [comment]
(Still only compile tested though)

File: platform/api/aws/images.go [hunk 1]
```diff
@@ -370,6 +370,43 @@ func (a *API) CreatePVImage(snapshotID string, diskSizeGiB uint, name string, de
...
+	}
+
+	snapshot, err := a.FindSnapshot(name)
+	if err != nil {
+		return err
+	}
+	if snapshot != nil {
+		// We explicitly ignore errors here in case somehow another AMI was based
+		// on that snapshot
+		a.ec2.DeleteSnapshot(&ec2.DeleteSnapshotInput{SnapshotId: &snapshot.SnapshotID})
```
Comment (arithx) [inline platform/api/aws/images.go]: Let's capture the error here and if it's not `nil` perform a `plog.Warnf("deleting snapshot %s: %v", snapshot.SnapshotID, err)`

**Comment by cgwalters** [comment]
OK actually testing this it didn't work because I was missing the `-hvm` for the `FindImage()` call.  Fixed that.  Added some logging too.

Also it's a bit tempting to just change the code and flip that to suffixing with `-pv` for the PV image and not suffixing what everyone uses now...but eh, separate PR.

**Comment by cgwalters** [comment]
See also https://github.com/coreos/coreos-assembler/pull/698

**Comment by cgwalters** [comment]
I don't have merge to this repo, what is the procedure?

**Comment by ajeddeloh** [comment]
Just say "ready for merge".

**Pull request merged**